modules/rostests/kmtests/rtl/RtlStrSafe.c

```diff
@@ -0,0 +1,179 @@
+    ok_eq_char(UsString.Buffer[5], L'4');
+    ok_eq_char(UsString.Buffer[6], (WCHAR) 0);
+
+    ///* STATUS_INVALID_PARAMETER tests */
+
+    ok_eq_hex(RtlUnicodeStringPrintf(NULL, FormatStringStrs, L"AAA", L"BBB", L"CCC"), STATUS_INVALID_PARAMETER);
+    ok_eq_hex(RtlUnicodeStringPrintf(&UsString, NULL, L"AAA", L"BBB", L"CCC"), STATUS_INVALID_PARAMETER);
+
+    UsStringNull.Buffer = (PWCH)NULL;
+    UsStringNull.Length = 0;
+    UsStringNull.MaximumLength = 0;
+    ok_eq_bool(RtlUnicodeStringPrintf(&UsStringNull, FormatStringStrs, L"AAA", L"BBB", L"CCC"), STATUS_INVALID_PARAMETER);
+
+    /* Test  for buffer overruns */
+
+    RtlFillMemory(Buffer, sizeof(Buffer), 0xAA);
+    RtlFillMemory(OvrBuffer, sizeof(OvrBuffer), 0xAA);
+    UsString.Buffer = Buffer;
+    UsString.Length = 0;
+    UsString.MaximumLength = 16 * sizeof(WCHAR);
+
+    ok_eq_hex(RtlUnicodeStringPrintf(&UsString, FormatStringStrs, L"abc", L"def", L"ghi"), STATUS_SUCCESS);
+    ok_eq_uint(UsString.Length, sizeof(L"abc def ghi") -sizeof(WCHAR));
+    ok_eq_char(UsString.Buffer[11], (WCHAR)0);
+    ok_eq_uint(0, memcmp(OvrBuffer + 12, Buffer + 12, sizeof(Buffer) - (12 * sizeof(WCHAR))));
+}
+
+TESTAPI
+Test_RtlUnicodeStringPrintfEx()
+{
+    WCHAR Buffer[32];
+    WCHAR BufferSmall[8] = { 0 };
+    WCHAR OvrBuffer[1024];
+    UNICODE_STRING UsString, RemString;
+    const WCHAR FormatStringInts[] = L"%d %d %d";
+    const WCHAR FormatStringStrs[] = L"%s %s %s";
+    const WCHAR Result[] = L"1 2 3";
+
+    /* No zeros (Don't assume UNICODE_STRINGS are NULL terminated) */
+
+    RtlFillMemory(Buffer, sizeof(Buffer), 0xAA);
+
+    UsString.Buffer = Buffer;
+    UsString.Length = 0;
+    UsString.MaximumLength = sizeof(Buffer);
+
+    RemString.Buffer = (PWCH)NULL;
+    RemString.Length = 0;
+    RemString.MaximumLength = 0;
+
+    /* STATUS_SUCCESS test, fill behind flag: low-byte as fill character */
+
+    ok_eq_hex(RtlUnicodeStringPrintfEx(&UsString, &RemString, STRSAFE_FILL_BEHIND | 0xFF, FormatStringInts, 1, 2, 3), STATUS_SUCCESS);
+    
+    ok_eq_uint(UsString.Length, sizeof(Result) - sizeof(WCHAR));
+    ok_eq_uint(UsString.MaximumLength, sizeof(Buffer));
+    ok_eq_uint(memcmp(UsString.Buffer, Result, sizeof(Result) - sizeof(WCHAR)), 0);
+
+    ok_eq_pointer(RemString.Buffer, UsString.Buffer + (UsString.Length / sizeof(WCHAR)));
+    ok_eq_uint(RemString.Length, 0);
+    ok_eq_uint(RemString.MaximumLength, UsString.MaximumLength - UsString.Length);
+
+    /* STATUS_BUFFER_OVERFLOW test  */
+
+    UsString.Buffer = BufferSmall;
+    UsString.Length = 0;
+    UsString.MaximumLength = sizeof(BufferSmall);
+
+    RemString.Buffer = (PWCH)NULL;
+    RemString.Length = 0;
+    RemString.MaximumLength = 0;
+    
+    ok_eq_hex(RtlUnicodeStringPrintfEx(&UsString, &RemString, 0, FormatStringStrs, L"AAA", L"BBB", L"CCC"), STATUS_BUFFER_OVERFLOW);
+    ok_eq_uint(UsString.Length, UsString.MaximumLength);
+    ok_eq_char(UsString.Buffer[0], L'A');
```

Comment:
Why not `memcmp`?

Reply:
This would require some `ExpectedBuffer` variable for each test to compare with. As we can neglect the performance penalty of not using `memcmp`, it's easier to just compare character by character here.
The added benefit of comparing each character is also that the test can tell you exactly which characters are right and which characters are wrong.